torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,229 @@
+from typing import List, Callable, Optional, Union, TypeVar, cast
+import torch.distributed as dist
+from .api import CheckpointException
+
+T = TypeVar('T')
+R = TypeVar('R')
+
+class _DistWrapper:
+    """
+    This is a wrapper around PG that provides a series of features around object collectives.
+
+    It works without distributed initialized, where most collectives turns into nops.
+
+    All variants that take functions are exception robust, meaning that if one or more
+    ranks raise errors, all ranks will observe those.
+    """
+    def __init__(self, group: Optional[dist.ProcessGroup], use_dist: bool, coordinator_rank: int):
```

Comment:
Is there is use case where we need to use rank other than 0 as coordinator_rank? Is it used for bcast or similar collective? Not sure if I understand correctly, is use_dist supposed to be True only in single process use case? Or there is a use case when the world_size is larger than 1, but use_dist is True?

Reply:
There are no know use-case for using a rank different than zero.
It's the sort of decision that we shouldn't hardcode at this layer since it's not capable of ensuring that's the right choice.

`use_dist == True` mean single process with distributed not initialized. This design comes from `save_state_dict` / `load_state_dict` which have that argument.

The decision was to make the user explicitly enable non-distributed to avoid mistakes of calling this API before initing a PG.